Content.Shared/_Sunrise/Weapons/Ranged/ConsumableAmmoSystem.cs

```diff
@@ -0,0 +1,134 @@
+    {
+        [Dependency] private readonly IEntityManager _entityManager = default!;
+        [Dependency] private readonly SharedPopupSystem _popupSystem = default!;
+        [Dependency] private readonly SharedStackSystem _stackSystem = default!;
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+        [Dependency] private readonly SharedProjectileSystem _projectileSystem = default!;
+        public override void Initialize()
+        {
+            base.Initialize();
+            SubscribeLocalEvent<ConsumableAmmoComponent, InteractUsingEvent>(OnLoadAmmo);
+            SubscribeLocalEvent<ConsumableAmmoComponent, ShotAttemptedEvent>(OnShotAttempted);
+            SubscribeLocalEvent<ConsumableAmmoComponent, TakeAmmoEvent>(OnTakeAmmo);
+            SubscribeLocalEvent<ConsumableAmmoComponent, ExaminedEvent>(OnExamine);
+        }
+
+        private void OnLoadAmmo(EntityUid uid, ConsumableAmmoComponent ammo, InteractUsingEvent args)
+        {
+            if (args.Handled)
+            {
+                return;
+            }
+
+            var itemInHand = args.Used;
+
+            if (!TryComp<StackComponent>(itemInHand, out var stack))
+            {
+                return;
+            }
+
+            var itemProtoId = _entityManager.GetComponent<MetaDataComponent>(itemInHand).EntityPrototype?.ID ?? string.Empty;
+
+            if (!ammo.LoadableItemIds.Contains(itemProtoId) && !ammo.MultiplerLoadableItemIds.Contains(itemProtoId))
+            {
+                return;
+            }
+
+            if (ammo.CurrentCharges >= ammo.MaxCharges)
+            {
+                _popupSystem.PopupPredicted(Loc.GetString("consumable-ammo-fully-charged"), uid, args.User);
+                args.Handled = true;
+                return;
+            }
+
+            var chargesPerItem = 1.0f / ammo.ItemsPerCharge;
+            if (ammo.MultiplerLoadableItemIds.Contains(itemProtoId))
+            {
+                chargesPerItem *= ammo.Multipler;
+            }
+
+            var chargesCanAdd = ammo.MaxCharges - ammo.CurrentCharges;
+            var itemsNeeded = (int)Math.Ceiling(chargesCanAdd / chargesPerItem);
+            var itemsToConsume = Math.Min(itemsNeeded, stack.Count);
+
+            if (itemsToConsume == 0)
+            {
+                return;
+            }
+
+            if (!_stackSystem.Use(itemInHand, itemsToConsume, stack))
+            {
+                return;
+            }
+
+            var actualChargesAdded = (int)Math.Floor(itemsToConsume * chargesPerItem);
+
+            var roomLeft = ammo.MaxCharges - ammo.CurrentCharges;
+            actualChargesAdded = Math.Min(actualChargesAdded, roomLeft);
+
+            ammo.CurrentCharges += actualChargesAdded;
+
+            _popupSystem.PopupPredicted(Loc.GetString("consumable-ammo-charged", ("chargesAdded", actualChargesAdded), ("currentCharges", ammo.CurrentCharges), ("maxCharges", ammo.MaxCharges)), uid, args.User);
```

Comment:
```suggestion
            
            var message = Loc.GetString("consumable-ammo-charged", ("chargesAdded", actualChargesAdded), ("currentCharges", ammo.CurrentCharges), ("maxCharges", ammo.MaxCharges));
            _popupSystem.PopupPredicted(message , uid, args.User);
```

Большие строки лучше создавать как отдельную переменную для читабельности